### ISSUE TITLE
Per-component runtime feature flags (re: 16.7.0 and hooks)

### ISSUE THREAD

**Issue by 1attice** [facebook/react#14531]
**Do you want to request a *feature* or report a *bug*?**

Feature

**What is the current behavior?**

When I did `npm-check-updates -u` I got randomly kicked to react 16.7.0 from 16.7.0-alpha.2. Hooks stopped working. 

The release notes mention that these features are behind a feature flag, but did not demonstrate how to set this flag, or link to a demonstration, indicating how this feature flag could be set. I've just trawled through the top ten Google hits for feature flags in React, and checked the documentation, and there is nothing. I'm assuming this is not a documentation oversight (I would have filed a ticket there instead, if I did) and will instead leap to the conclusion that this functionality does not yet exist. The feature flag is presumably something set during packaging, and I can't rely on being able to set it programmatically from a stock 16.7.0.

This is actually a big deal for me, as I've been using hooks extensively in a large codebase for an as-of-yet-unreleased project. My understanding was that they'd show up in non-alpha within a month or two (when I began moving over from recompose's `withState`, etc.) Now, I'm going through and removing all the hooks. I feel dismayed. It will be a while before I bother with them again. 

On the other hand, I respect & agree with the reasoning presented; that React's stability is paramount, and that it is wrong to release features that change developer behaviour in ways that are not yet fully appreciated. 

Generalizing, this is a problem that is likely to recur as React continues its swift, admirable ascent. So it goes on the frontend. And while I don't think the React team is going to make the same misstep that Angular did (1->2), I think there is an emergent, generalized tension in this project between improving things and the risk of breaking things. Which, upon reflection, is considerable, given (a) how popular React has become, and (b) how rapidly the frontend space changes. 

Generalized problems call for generalized solutions. So, I'm going to propose a general-purpose means of setting feature flags per component. If this already exists, please let me know :3 Otherwise, please accept this proposal for, let's call them, React Feature Flags.

**What is the expected behavior?**

If there are features, available via flags, these flags should be settable at runtime. To me, the place that makes the most sense would be setting a property on a given component, as rendered: 
```
import React, { useSomeHook } from 'react'

const Foo = ({ children }) => <div>{children}</div>

Foo.features = { hooks: true } // rhymes with how defaultProps are currently set on pure functional components

const Bar = () => {
   const [qux] = useSomeHook()
   // ...
   return <div>{qux}</div>
}

```

All child elements of Foo, and Foo itself, would thus have hooks enabled, and could be called like `<Foo><Bar /></Foo>`

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

All versions. 

I'm imagining as features become tried-and-true they move from being disabled by default to being enabled by default; after a significant upgrade to React, everything would keep working, but there might be places where the flags could be removed without affecting operation. 

What's cool about my proposal, I think, is that it also allows code using new/alpha features to constrain the area of effect for those features, isolating any potential breakage to just those components that need the new-shiny. 

Or, if one is lazy or doing a quick-and-dirty, you could just set the flag once on the top-level component for your SLA, and have all the features you want, everywhere, with maximum concision.

**Comment by aweary** [comment]
> Now, I'm going through and removing all the hooks. I feel dismayed. It will be a while before I bother with them again.

Hooks are still going to be published in a stable release at some point. The fact that 16.7.0 was published without Hooks does not imply that the React team is changing course at all. Please see [this blog post](https://reactjs.org/blog/2018/12/19/react-v-16-7.html) for details on this release and why this was released with Hooks.

If you'd like to keep using Hooks, please continue doing so! You just need to pin your dependencies to `16.7.0-alpha.2` (no `^`) for now. See https://github.com/facebook/react/issues/14481 for some more context.

As far as per-component feature flags go, I don't think it's an API we'd want to implement. React would need to check every component's feature flags before performing any actions that could be affected by feature flags, which would potentially add significant overhead.

The API is also too granular, since it's unlikely that anyone who is using a new API would only be doing so in a small number of components. For example, if I wanted to use Hooks in every component for my new side project, I'd have to add a `features` object to every single one (which also negatively affects code size). It's also problematic for features that aren't typically isolated to a single component, like context or concurrent rendering. There are some other reasons it's not ideal, but hopefully that was sufficient.

*tl;dr* Feature flags are tricky, Hooks are still coming, pin the `16.2.0-alpha.2` release to keep using them for now.

Thanks!
